pallets/dca/src/lib.rs

```diff
@@ -0,0 +1,1029 @@
+	#[pallet::storage]
+	#[pallet::getter(fn schedules)]
+	pub type Schedules<T: Config> =
+		StorageMap<_, Blake2_128Concat, ScheduleId, Schedule<T::AccountId, T::Asset, BlockNumberFor<T>>, OptionQuery>;
+
+	/// Storing schedule ownership
+	#[pallet::storage]
+	#[pallet::getter(fn owner_of)]
+	pub type ScheduleOwnership<T: Config> =
+		StorageDoubleMap<_, Blake2_128Concat, T::AccountId, Twox64Concat, ScheduleId, (), OptionQuery>;
+
+	/// Keep tracking the remaining amounts to spend for DCA schedules
+	#[pallet::storage]
+	#[pallet::getter(fn remaining_amounts)]
+	pub type RemainingAmounts<T: Config> = StorageMap<_, Blake2_128Concat, ScheduleId, Balance, OptionQuery>;
+
+	/// Keep tracking the retry on error flag for DCA schedules
+	#[pallet::storage]
+	#[pallet::getter(fn retries_on_error)]
+	pub type RetriesOnError<T: Config> = StorageMap<_, Blake2_128Concat, ScheduleId, u32, OptionQuery>;
+
+	/// Keep tracking of the schedule ids to be executed in the block
+	#[pallet::storage]
+	#[pallet::getter(fn schedule_ids_per_block)]
+	pub type ScheduleIdsPerBlock<T: Config> =
+		StorageMap<_, Blake2_128Concat, BlockNumberFor<T>, BoundedVec<ScheduleId, T::MaxSchedulePerBlock>, ValueQuery>;
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T>
+	where
+		<T as pallet_route_executor::Config>::AssetId: From<<T as pallet::Config>::Asset>,
+		<T as pallet_route_executor::Config>::Balance: From<u128>,
+		u128: From<<T as pallet_route_executor::Config>::Balance>,
+	{
+		/// Creates a new DCA schedule and plans the execution in the specified start execution block.
+		/// If start execution block number is not specified, then the schedule is planned in the consequent block.
+		///
+		/// The order will be executed within the configured AMM trade pool
+		///
+		/// Parameters:
+		/// - `origin`: schedule owner
+		/// - `schedule`: schedule details
+		/// - `start_execution_block`: start execution block for the schedule
+		///
+		/// Emits `Scheduled` and `ExecutionPlanned` event when successful.
+		///
+		#[pallet::call_index(0)]
+		#[pallet::weight(<T as Config>::WeightInfo::schedule())]
+		#[transactional]
+		pub fn schedule(
+			origin: OriginFor<T>,
+			schedule: Schedule<T::AccountId, T::Asset, BlockNumberFor<T>>,
+			start_execution_block: Option<BlockNumberFor<T>>,
+		) -> DispatchResult {
+			let who = ensure_signed(origin.clone())?;
+			ensure!(who == schedule.owner, Error::<T>::Forbidden);
+
+			ensure!(schedule.order.get_route_length() > 0, Error::<T>::RouteNotSpecified);
+
+			let min_budget = if schedule.order.get_asset_in() == T::NativeAssetId::get() {
+				T::MinBudgetInNativeCurrency::get()
+			} else {
+				Self::convert_native_amount_to_currency(
+					schedule.order.get_asset_in(),
+					T::MinBudgetInNativeCurrency::get(),
+				)?
+			};
+			ensure!(
+				schedule.total_amount >= min_budget,
+				Error::<T>::TotalAmountIsSmallerThanMinBudget
+			);
+
+			let weight_for_single_execution = <T as Config>::WeightInfo::on_initialize_with_one_trade();
```

Comment:
based on benchmarks you'll probably have to charge fee based on number of trades in route and type of the tx (sell/buy)

Reply:
we discussed in a call some weeks ago that we will use buy for both sell and buy, so be optimistic with it.

But I can benchmark it separately for buy and sell.

As for the number of trades, we will have only one as we have only omnipool AMM in hydra atm.

Reply:
we will have only one route now, but that will probably have to change when we add additional AMM's to the runtime. i would leave some todo in the runtime so we don't forget to update this. 

i would definitely benchmark both buy and sell and based on it we can decide what will be charged as user can do both currently